File: relationship-manager/conversation-service/src/main/java/com/bank/rm/conversation/engine/ConversationStateMachine.java
package com.bank.rm.conversation.engine;

import com.bank.rm.common.dto.ConversationPhase;
import java.util.Map;
import java.util.Set;

public class ConversationStateMachine {

    private static final Map<ConversationPhase, Set<String>> PHASE_REQUIREMENTS = Map.of(
        ConversationPhase.GREETING, Set.of("name"),
        ConversationPhase.PERSONAL, Set.of("ageGroup", "location"),
        ConversationPhase.FINANCIAL, Set.of("incomeSource", "incomeRange"),
        ConversationPhase.GOALS, Set.of("retirementTarget"),
        ConversationPhase.RISK_ASSESSMENT, Set.of("riskTolerance"),
        ConversationPhase.RECOMMENDATION, Set.of(),
        ConversationPhase.CHANNEL_PREF, Set.of("preferredChannel"),
        ConversationPhase.FOLLOWUP_SCHEDULE, Set.of("followUpFrequency")
    );

    public boolean canTransition(ConversationPhase currentPhase, Map<String, Object> extractedData) {
        Set<String> required = PHASE_REQUIREMENTS.getOrDefault(currentPhase, Set.of());
        return required.stream().allMatch(key -> extractedData.containsKey(key) &&
                extractedData.get(key) != null &&
                !extractedData.get(key).toString().isEmpty());
    }

    public ConversationPhase getNextPhase(ConversationPhase currentPhase) {
        return switch (currentPhase) {
            case GREETING -> ConversationPhase.PERSONAL;
            case PERSONAL -> ConversationPhase.FINANCIAL;
            case FINANCIAL -> ConversationPhase.GOALS;
            case GOALS -> ConversationPhase.RISK_ASSESSMENT;
            case RISK_ASSESSMENT -> ConversationPhase.RECOMMENDATION;
            case RECOMMENDATION -> ConversationPhase.CHANNEL_PREF;
            case CHANNEL_PREF -> ConversationPhase.FOLLOWUP_SCHEDULE;
            case FOLLOWUP_SCHEDULE -> ConversationPhase.COMPLETED;
            case COMPLETED -> ConversationPhase.COMPLETED;
        };
    }
}

File: relationship-manager/conversation-service/src/main/kotlin/com/bank/rm/conversation/engine/ConversationStateMachine.kt
package com.bank.rm.conversation.engine

import com.bank.rm.common.dto.ConversationPhase
import com.bank.rm.common.dto.ConversationStatus
import org.springframework.stereotype.Component

@Component
class ConversationStateMachine {

    private val validTransitions = mapOf(
        ConversationPhase.GREETING to setOf(ConversationPhase.PERSONAL),
        ConversationPhase.PERSONAL to setOf(ConversationPhase.FINANCIAL),
        ConversationPhase.FINANCIAL to setOf(ConversationPhase.GOALS),
        ConversationPhase.GOALS to setOf(ConversationPhase.RISK_ASSESSMENT),
        ConversationPhase.RISK_ASSESSMENT to setOf(ConversationPhase.RECOMMENDATION),
        ConversationPhase.RECOMMENDATION to setOf(ConversationPhase.CHANNEL_PREF),
        ConversationPhase.CHANNEL_PREF to setOf(ConversationPhase.FOLLOWUP_SCHEDULE),
        ConversationPhase.FOLLOWUP_SCHEDULE to setOf(ConversationPhase.COMPLETED)
    )

    fun canTransition(from: ConversationPhase, to: ConversationPhase): Boolean =
        validTransitions[from]?.contains(to) == true

    fun getNextPhase(current: ConversationPhase): ConversationPhase? =
        validTransitions[current]?.firstOrNull()

    fun isTerminal(phase: ConversationPhase): Boolean =
        phase == ConversationPhase.COMPLETED

    fun determineStatus(phase: ConversationPhase): ConversationStatus = when (phase) {
        ConversationPhase.COMPLETED -> ConversationStatus.COMPLETED
        else -> ConversationStatus.ACTIVE
    }
}
